### PR TITLE
feat/find-trip-id-for-cancellations

### PR DESCRIPTION
**Pull request by mjaakko**
This is a workaround needed to display trip cancellations in Google

### CODE REVIEW

File: src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            return tripIdCache.get(new TripDetails(routeId, operatingDay, startTime, directionId));
+        } catch (ExecutionException e) {
+            logger.warn("Failed to fetch trip id for {} / {} / {} / {}", routeId, operatingDay, startTime, directionId, e);
+            return Optional.empty();
+        }
+    }
+
+    private static Optional<String> getTripIdFromDigitransitAPI(String routeId, String operatingDay, String startTime, int directionId) throws IOException {
+        //TODO: maybe this URL should be configurable?
+        URLConnection connection = new URL("https://api.digitransit.fi/routing/v1/routers/hsl/index/graphql").openConnection();
```
Comment (Divinare) [inline src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java]: Should this url could come from environment variables? We define it at stack description .yml file and hardcode it inside project's environment.conf

File: src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        connection.setDoOutput(true);
+        connection.setRequestProperty("Content-Type", "application/graphql");
+        connection.connect();
+
+        try (OutputStream os = connection.getOutputStream()) {
+            os.write(String.format(FUZZY_TRIP_QUERY, routeId, PubtransFactory.joreDirectionToGtfsDirection(directionId), operatingDay, startTimeToSeconds(startTime)).getBytes(StandardCharsets.UTF_8));
+        }
```
Comment (Divinare) [inline src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java]: catch and log something like "Failed to connect digitransitRouteAPI"

File: src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+        try (InputStream is = connection.getInputStream()) {
```
Comment (Divinare) [inline src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java]: catch and log something like "Failed to read input stream from digitransitRouteAPI

File: src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                if (data.has("fuzzyTrip")) {
+                    JSONObject fuzzyTrip = data.optJSONObject("fuzzyTrip");
+                    if (fuzzyTrip.has("gtfsId")) {
+                        return Optional.ofNullable(fuzzyTrip.optString("gtfsId"));
+                    }
+                }
+            }
+
+            return Optional.empty();
+        }
```
Comment (Divinare) [inline src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java]: I would add a line:
return Optional.empty();

File: src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+ */
+public class TripIdCache {
+    private static final Logger logger = LoggerFactory.getLogger(TripIdCache.class);
+
+    private static final String FUZZY_TRIP_QUERY = "{ fuzzyTrip(route: \"HSL:%s\", direction: %d, date: \"%s\", time: %d) { gtfsId } }";
+
+    private LoadingCache<TripDetails, Optional<String>> tripIdCache = CacheBuilder.newBuilder().expireAfterAccess(1, TimeUnit.HOURS).build(new CacheLoader<TripDetails, Optional<String>>() {
+        @Override
+        public Optional<String> load(TripDetails tripDetails) throws Exception {
+            return getTripIdFromDigitransitAPI(tripDetails.routeId, tripDetails.operatingDay, tripDetails.startTime, tripDetails.directionId);
```
Comment (Divinare) [inline src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java]:
```
Optional<String> tripId = getTripIdFromDigitransitAPI(.....
if (tripId.isPresent()) {
       return tripId;
} else {
     log.warn("Failed to generate tripId for cancellation);
     return Optional.empty()
}
```

File: src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java (hunk 1 above)
Comment (mjaakko) [reply src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java]: Fixed

File: src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java (hunk 5 above)
Comment (mjaakko) [reply src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java]: Fixed

**Pull request merged**